packages/dds/tree/src/util/rangeMap.ts

```diff
@@ -147,3 +147,78 @@ export function setInRangeMap<T>(map: RangeMap<T>, start: number, length: number
 	const numContainedEntries = iAfter - iBefore - 1;
 	map.splice(iBefore + 1, numContainedEntries, newEntry);
 }
+
+/**
+ * Delete the keys from `start` to `start + length - 1`
+ *
+ * 1. If an entry is completely included in the deletion range, the whole entry will be deleted
+ * e.g.: map = [[1, 2], [4, 6]], delete range: [3, 6]
+ * map becomes [[1, 2]] after deletion
+ * (Note: the notation [a, b] represents start = a, end = b for simpler visiualization, instead of `b`
+ * representing the length)
+ *
+ * 2. If an entry is partially overlapped with the deletion range, the start or end point will be shifted
+ * e.g.: map = [[1, 2], [4, 6]], delete range: [2, 4]
+ * map becomes [[1, 1], [5, 6]] after deletion
+ *
+ * 3. If an entry completely includes the deletion range, the original entry may be split into two.
+ * e.g.: map = [[1, 6]], delete range: [2, 4]
+ * map becomes [[1, 1], [5, 6]]
+ */
+export function deleteFromRangeMap<T>(map: RangeMap<T>, start: number, length: number): void {
```

Comment:
I know you didn't add it in this PR, but I think it would be great to improve the docs for RangeMap.

The existing doc comment on RangeMap says "A map keyed on integers allowing reading and writing contiguous ranges of integer keys." I find that confusing since its not a map, and while it does have integer keys (since its an array), indexing into it doesn't actually do the desired lookup operation. Instead, you have to use getFromRangeMap whis is not mentioned form the docs.

RangeMap's docs should probably document its actual data invariants (I think it requires that its range entries are sorted by start and non-overlapping, but I'm not sure thats how it works: documenting that would help. I also am unclear on if it permits adjacent equal valued ranges or requires them to be merged) as well as what function to use to do the operations its documentation implies exist (ex: it says its a map, so it should document how to use it as a map since the type itself doesn't provide the needed functionality directly like actual Map types do).

It would also be great if rangeMap's docs were detailed enough that I could tell what delete means. I assume that a range map is logically a mapping from non-negative integer to `T | undefined` (docs should make that clear) and delete sets a range to undefined. If that is the case, maybe a better API would be to allow undefined to be passed into set?

Consider the case where T permits undefined, for example, RangeMap<unknown>. If using such a rangemap and I do setInRangeMap(m, 0, 1, undefined), then getFromRangeMap(m, 0, 1), I have no way to tell the difference between no value being set in the map and an explicit undefined. Givin this lack of distinction, I think documenting setting undefined to be the same as delete fixes this edge case.


Also the docs on getFromRangeMap don't say how the length passed into it is used (and the docs on RangeQueryResult don't cover how the passed in length impacts the returned length)

Reply:
So in short, I guess:

1. Range map needs better docs
2. if RangeMap had better docs, it would be much easier to review this.
3. We should think about how to handle explicit undefined values. If we want to support them the current API may need some tweaks, and if not it might make sense to unify delete and set and delete by setting values to undefined (we likely still want a separate delete function at least as an implementation detail, but set would call it in the undefined value case)

Reply:
Created tasks to address these [AB#7413](https://dev.azure.com/fluidframework/internal/_workitems/edit/7413), [AB#7414](https://dev.azure.com/fluidframework/internal/_workitems/edit/7414)